src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.StackRowStack.cs

```diff
@@ -11,18 +11,27 @@ public sealed partial class JsonDocument
     {
         private struct StackRowStack : IDisposable
         {
-            private byte[] _rentedBuffer;
+            private byte[]? _rentedBuffer;
             private int _topOfStack;
 
             internal StackRowStack(int initialSize)
             {
-                _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialSize);
-                _topOfStack = _rentedBuffer.Length;
+                if (initialSize == -1)
+                {
+                    // Used for primitive values that don't need a stack.
```

Comment:
Why is this? Do we have this same behavior even with `ParseUnrented`.

Reply:
This avoids using the pool below:
```cs
_rentedBuffer = ArrayPool<byte>.Shared.Rent(initialSize);
```

`ParseUnrented` passes -1 for primitive tokens (meaning not StartArray and StartObject tokens), otherwise a size is passed.

Note this optimization could also be applied for non-`ParseUnrented` methods (the "normal" flow without the serializer-only path) with a bit of refactoring around detecting the current token type.

Reply:
Note I updated this code to also apply the optimization to the "normal" flow. However, I can't measure perf completely until we get https://github.com/dotnet/performance/pull/852 in.

Reply:
I am trying to recall if there is any drawback or side effect of not adding a "row" for a particular JSON token into the metadata DB. At a quick glance, there doesn't seem to be an issue, but the mental model had always been "add a row for every token". Since most payloads won't be single token primitive values, in the context of `JsonDocument.Parse` alone, this PR is adding a bit of complexity to that.

However, since `JsonDocument.Parse` currently gets called a lot by the `JsonSerializer` while deserializing object properties, possibly with a small slice of the JSON corresponding to a single primitive token, I see why such a change would be super useful.

Something to verify: Are we changing any of the allocation characteristic/memory footprint for such cases (using GC-backed arrays rather than rented ones)?

Reply:
There is still a "row" created for every token; the code mentioned above is the `StackRowStack` which is used to update the number of rows or elements when the object\array is done being parsed. So this optimization avoids a pooled alloc within `StackRowStack` when the element being returned is the one and only element and also not an object\array.

Reply:
Oh... Nice!

That makes a lot more sense.